### PR TITLE
New feature: sw filters packets sampling

### PR DESCRIPTION
**Pull request by akadoorie**
I’ve added a new feature that I would like to share.
This new pf_ring feature allows packets sampling of packets that belong to a sw
filtered session. In other words, if a sw filter is applied, you will be able not to block
100% of the packets (till filtering rule removal). Some packets will be
allowed to pass through the filter.

The use case is keeping the advantage of improved performance by filtering
most (e.g. 90%) of the traffic, while being able to get some (e.g. 10%) of
the traffic for analysis purposes. This will allow the application not be
100% blind to a filtered session.

This feature is not good for applications that are sensitive to
out-of-sequence packets.

Technical code details:
-----------------------
- New pf_ring param “filtering_sampling_segment_size” is set to 100
by default.

- New pf_ring userspace API
int pfring_set_sw_filtering_sampling_rate(pfring ring, u_int32_t rate / 0 =
no sampling */) was added.

This API should be used right after adding a new SW hash filtering rule,
with a “rate” parameter set.
A rate=10 means that the 10 first packets of every segment of 100 filtered
packets of the session, will be forwarded, while the rest 90 will be filtered.

If the API is not called, then the default rate is 0, thus no change in
current behavior (i.e. 100% of the packets are filtered, since 0 filtered
packets are forwarded).

I’ve had to add a new member variable to struct pf_ring_socket
which is ‘sw_filtering_hash_filtered’ that counts the actual filtered
packets.

Current ‘sw_filtering_hash_match’ is not good enough anymore for this
purpose since it counts the matches before excluding the forwarded packets,
but it has a great help in calculating the sampled filtered packets.

Technical build/test details:
---------------------------
Functionality was tested based on pf_ring version 6.6.0 (on CentOS 7)

Any comments are welcome.

Amir

### CODE REVIEW

**Comment by cardigliano** [comment]
Thank you, we will review it asap!

**Comment by akadoorie** [comment]
Hi Alfredo,

I think there is a problem with the implementation, that I would like to consult with you:
Since the calculation is per rule rather than per ring, I need to change the "pfr->sw_filtering_hash_match % filtering_sampling_segment_size" part to something like "hash_bucket->match % filtering_sampling_segment_size". (This BTW, requires bringing hash_bucket from check_perfect_rules(..)). When I did this, I also wanted to improve the rule statistics, by adding a "filtered" member to struct sw_filtering_hash_bucket, just like "match" and increase it next to "pfr->sw_filtering_hash_filtered++".
When I tried to calculate how many packets where sampled (hash_bucket->match - hash_bucket->filtered) I saw that the numbers not always perfect (sometimes there are more filtered packets than matched packets on the same rule) which doesn't make sense.
Then it hit me that the code is not thread-safe, thus I need to add a lock.
I used write_lock_bh(&pfr->ring_rules_lock) and write_unlock_bh(&pfr->ring_rules_lock) to wrap the whole section that handle sw hash rules match and statistics update (i.e. where check_perfect_rules() is called) and the numbers began to make sense.

So now I have several questions that I hope you can assist me to figure out:
1. I see that the function check_wildcard_rules(..) is thread-safe while check_perfect_rules(..) doesn't. Does it make sense?
2. Assuming I use write_lock_bh(&pfr->ring_rules_lock) and write_unlock_bh(&pfr->ring_rules_lock), I'm afraid it could be very heavy inside add_skb_to_ring(..) in the place I described, since they are called per packet. What do you think about it?

**Comment by cardigliano** [comment]
Hi @akaduri75, a couple of comments:
1. it seems check_perfect_rules() is missing a read_lock() (not that heavy), to prevent concurrent accesses when a rule is added with SO_ADD_FILTERING_RULE (there is already a write_lock_bh() there)
2. since an interface can have multiple queues, you can have concurrent updates to sw_filtering_hash_filtered, thus sw_filtering_hash_filtered should be an atomic counter (another option is using a write_lock() per rule, but an atomic is better)

**Comment by akadoorie** [comment]
Thanks for the remarks Alfredo.
Regarding remark (1), I tried read locks, but it seems that the counters work properly only under write lock.
I suggest that I'll update my changes to the forked branch so you can see the final code, and comment again.
Amir

**Comment by cardigliano** [comment]
Correct, you are updating a counter, thus you need a write lock there.

**Comment by akadoorie** [comment]
Hi Alfredo,

I'll appreciate if we could make a progress with this.
AFAIU, the write_lock is for every packet, which can cause a some performance degradation. This is the only thing that I don't like in this implementation, but I don't see how to avoid it.

Thanks,
Amir

**Comment by cardigliano** [comment]
@akaduri75 I need some time to analyse and test this code, in my understanding a read_lock() (no _bh) and atomic counters should work fine, but you said it looks like a write_lock (_bh?) is needed

**Comment by akadoorie** [comment]
Correct.
When using a lot of traffic, I've noticed that the calculation "hfrs.miss = bucket->match - bucket->filtered" (that I do in getsockopt() in case SO_GET_HASH_FILTERING_RULE_STATS) gives very high values (which are actually negative but in an unsigned variable) because bucket->match < bucket->filtered, and this is "fixed" when I use write_lock_bh() only (in add_skb_to_ring()). read_lock() probably not enough.
The dilema is that that I do want these statistics corrected, but on the other hand I don't want write_lock on every packet just for statistics purposes.
Anyway, if you put aside the statistics issue, the most important fix here is when I use hash_bucket->match (instead of pfr->sw_filtering_hash_match) to do the modulo (%) calculation, for the new filtering sampling feature.

**Comment by akadoorie** [comment]
One correction:
Of course, in order to use hash_bucket->match inside add_skb_to_ring(), the pointer reference for hash_backet, from check_perfect_rules() up to add_skb_to_ring() is also required (as can be seen in my code changes).

**Comment by akadoorie** [comment]
I'll appreciate reconsidering this feature.
I can add that I use this feature, together with the filtering statistics enhancements (all committed in the branch of course) for more than 2 months and I benefit a lot from it: I both have filters that don't 100% hermetic, and more accurate rule statistics.
I'm using the code in this branch, with a slight changed: I commented out the locks. I.e. lines 3484 and 3505 are commented out. Also, in line 3498, I increment pfr->sw_filtering_hash_filtered with ++ (i.e. unsafe) instead of atomic64_inc(). The reason is that I was worried that the locks will cause significant performance degradation. Since the sampling feature is statistical, I preferred to live with degraded accuracy than degraded performance. The result is, as I mentioned on Nov 19, 2017, that on low sampling rate (0 or close to 0), hfrs.miss (in my branch) is very high (negative actually). Its maybe not nice, but I can live with it. (Obviously, the locks solve this issue).
Bottom line, if the locks are not as bad as I think (performance-wise), its possible to keep the code as is. Otherwise, its possible to remove them.

**Comment by akadoorie** [comment]
PR https://github.com/ntop/PF_RING/pull/350 replaced this one.
